### PR TITLE
travis: Use g++-7 direcly from packages (and fix ubuntu 16.04 problems)

### PR DESCRIPTION
**Pull request by marehr**

### CODE REVIEW

**Comment by h-2** [comment]
Whats going on here :question:

**Comment by marehr** [comment]
I try to use the gcc-7 compiler from https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test, because our current compiler does weird things. Apparently the native gcc-7 compiler has some bug that I try to workaround/fix :(

**Comment by h-2** [comment]
What is the status here?

**Comment by marehr** [comment]
@rrahn mergeable

File: include/seqan3/range/container/concept.hpp [hunk 1]
```diff
@@ -42,6 +42,50 @@
...
 #include <initializer_list>
 
+// remove if sequence_concept_modified_by_const_iterator_bug vanished from travis
+#include <string>
+
+namespace seqan3::detail
+{
+//!\privatesection
+
+/*!\interface seqan3::sequence_concept_modified_by_const_iterator <>
```
Comment (rrahn) [inline include/seqan3/range/container/concept.hpp]: whitespace in docu.

File: include/seqan3/range/container/concept.hpp [hunk 2]
```diff
@@ -127,15 +171,21 @@ concept bool sequence_concept = requires (type val, type val2)
...
+    // because of a travis bug we can't assume typename type::iterator as return type
+    { val.insert(val.begin(), typename type::size_type{}, typename type::value_type{}) };
+    { val.insert(val.begin(), val2.begin(), val2.end())                                };
+    //TODO should return type::iterator on strings (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83328)
+    { val.insert(val.begin(), std::initializer_list<typename type::value_type>{})      };
+    { val.erase(val.begin())                                                           } -> typename type::iterator;
+    { val.erase(val.begin(), val.end())                                                } -> typename type::iterator;
+
+    // workaround a travis bug where insert/erase can't take a const iterator, e.g. cbegin()
+    requires detail::sequence_concept_modified_by_const_iterator_bug || detail::sequence_concept_modified_by_const_iterator<type>;
```
Comment (rrahn) [inline include/seqan3/range/container/concept.hpp]: Please add a check, that this is only tested if the queried type is in fact a std::basic_string.

File: include/seqan3/range/container/concept.hpp [hunk 3]
```diff
@@ -42,6 +42,50 @@
...
+ *
+ * seqan3::detail::sequence_concept_modified_by_const_iterator<std::string> is
+ * known to work, but travis-ci has version of g++-7.2 where a bug in the
+ * STL prevents this concept to be true.
+ *
+ * \attention This workaround can be removed if container_concept_test is not
+ * failing on travis anymore.
+ */
+constexpr bool sequence_concept_modified_by_const_iterator_bug
+    = sequence_concept_modified_by_const_iterator<std::string> == false;
```
Comment (rrahn) [inline include/seqan3/range/container/concept.hpp]: Can we not use basic_string? Wouldn't that be safer?

File: include/seqan3/range/container/concept.hpp [hunk 4]
```diff
@@ -42,6 +42,50 @@
...
 
+// remove if sequence_concept_modified_by_const_iterator_bug vanished from travis
+#include <string>
+
+namespace seqan3::detail
+{
+//!\privatesection
+
+/*!\interface seqan3::sequence_concept_modified_by_const_iterator <>
+ * \brief Can insert and erase used with const iterators
```
Comment (rrahn) [inline include/seqan3/range/container/concept.hpp]: Maybe: `Checks whether insert and erase can be used with const_iterator`.
And maybe also add an explantation why this was separated from the original concept definition,
and that it should go back in, if the travis bug is not of any concern anymore.

**Pull request merged**